tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+            let j_idx = (j + head) as usize & MASK;
+
+            // Get the next pointer
+            let next = if j == n {
+                // The last task in the local queue being moved
+                task.header().into()
+            } else {
+                self.inner.buffer[j_idx].with(|ptr| unsafe {
+                    let value = (*ptr).as_ptr();
+                    (*value).header().into()
+                })
+            };
+
+            self.inner.buffer[i_idx].with_mut(|ptr| unsafe {
+                let ptr = (*ptr).as_ptr();
+                *(*ptr).header().queue_next.get() = Some(next);
+            });
+        }
+
+        let head = self.inner.buffer[head as usize & MASK]
+            .with(|ptr| unsafe { ptr::read((*ptr).as_ptr()) });
+
+        // Push the tasks onto the inject queue
+        inject.push_batch(head, task, BATCH_LEN);
+
+        Ok(())
+    }
+
+    /// Pops a task from the local queue.
+    pub(super) fn pop(&mut self) -> Option<task::Notified<T>> {
+        // If a task is available in the FIFO slot, return that.
+        if let Some(task) = self.next.take() {
+            return Some(task);
+        }
+
+        loop {
+            let head = self.inner.head.load(Acquire);
+
+            // safety: this is the **only** thread that updates this cell.
+            let tail = unsafe { self.inner.tail.unsync_load() };
+
+            if head == tail {
+                // queue is empty
+                return None;
+            }
+
+            // Map the head position to a slot index.
+            let idx = head as usize & MASK;
+
+            let task = self.inner.buffer[idx].with(|ptr| {
+                // Tentatively read the task at the head position. Note that we
+                // have not yet claimed the task.
+                //
+                unsafe { ptr::read(ptr) }
+            });
+
+            // Attempt to claim the task read above.
+            let actual = self
+                .inner
+                .head
+                .compare_and_swap(head, head.wrapping_add(1), Release);
+
+            if actual == head {
+                return Some(unsafe { task.assume_init() });
+            }
+
+            atomic::spin_loop_hint();
+        }
+    }
+}
+
+impl<T> Steal<T> {
+    pub(super) fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Steals half the tasks from self and place them into `dst`.
+    pub(super) fn steal_into(&self, dst: &mut Local<T>) -> Option<task::Notified<T>> {
+        let dst_tail = unsafe { dst.inner.tail.unsync_load() };
+
+        // Steal the tasks into `dst`'s buffer. This does not yet expose the
+        // tasks in `dst`.
+        let mut n = self.steal_into2(dst, dst_tail);
+
+        if n == 0 {
+            // No tasks were stolen
+            return None;
+        }
+
+        // We are returning a task here
+        n -= 1;
+
+        let ret_pos = dst_tail.wrapping_add(n);
```

Comment:
This means we end up executing the last element in FIFO order from tasks we steal. Not sure if this matters in practice, but wouldn't we prefer to execute the head-most task we stole?

Reply:
It's not obvious it can trivially be done. The producer only has the ability to touch the tail of the queue w/o synchronization. In order to pop the oldest task, the queue would need to be fully synchronized, which would make the tasks that were **just** stolen available to other stealers, so it would be possible to steal tasks then not have a task to return...

Given that this is how it works today, i would punt for further investigation.